pennylane/ops/op_math/composite.py

```diff
@@ -135,9 +137,38 @@ def has_matrix(self):
     def eigvals(self):
         """Return the eigenvalues of the specified operator."""
 
+    def _get_cached_matrix(self, wire_order=None, cache=False):
+        """Contains the caching logic for the matrix representation of the operator."""
+        wire_order = wire_order or self.wires
+        wire_order_hash = hash(Wires(wire_order))
+        base_wire_hash = hash(self.wires)
+
+        if wire_order_hash in self._mat_cache:  # result already cached
+            return self._mat_cache[wire_order_hash]
+
+        if base_wire_hash in self._mat_cache:  # result not in cache, but base in cache
+            res = math.expand_matrix(self._mat_cache[base_wire_hash], self.wires, wire_order)
+
+        else:
+            res = self.compute_matrix(
+                self.wires,
+                wire_order,
+                operands=self.operands,
+                has_overlapping_wires=self.has_overlapping_wires,
+            )
+
+        if cache:
+            self._mat_cache[wire_order_hash] = res  # set cache
+        return res
+
+    @staticmethod
     @abc.abstractmethod
-    def matrix(self, wire_order=None):
+    def compute_matrix(*args, **kwargs):
```

Comment:
I don't like using `compute_matrix` for this method. As mentioned in the `Operator.compute_matrix` docstring:
```
Representation of the operator as a canonical matrix in the computational basis (static method).

The canonical matrix is the textbook matrix representation that does not consider wires.
Implicitly, this assumes that the wires of the operator correspond to the global wire order.
```
But in here we are considering the wires...

I would just change this method's name (to `_matrix` for example) and also remove the `@staticmethod`.

Reply:
I don't mind getting rid of `compute_matrix` altogether and letting it resolve to `Operator.compute_matrix()` if someone calls it (which would raise a `MatrixUndefinedError`). Perhaps in this case we should override `has_matrix()` to be `True` despite `Operator.has_matrix()` stating otherwise?

Actually-let me ask in a different way. What is the purpose of `matrix()` vs. `has_matrix()` vs `compute_matrix()`? Instances of this class have matrices defined, but the class itself inherently does not, which is fine!

Reply:
In the past the need to determine if an operation has a matrix would require explicitly building the matrix itself. This can be time consuming for large operations. For this reason `has_matrix()` was introduced to keep track of this quantity without having to explicitly compute the matrix. 

@AlbertMitjans I think the `compute_matrix()` is designed to be static. That being said, I agree that we are kind of abusing it here by passing in wires. I will update the logic to reflect this 👍🏼